lib/src/main/java/io/stormbird/token/tools/TokenScriptValidator.java

```diff
@@ -0,0 +1,250 @@
+        TrustManagerFactory tmf = TrustManagerFactory
+                .getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        // 'null' will initialise the tmf with the default CA certs installed
+        // with the JRE.
+        tmf.init((KeyStore) null);
+        X509TrustManager tm = (X509TrustManager) tmf.getTrustManagers()[0];
+        X509Certificate[] validIssuers = tm.getAcceptedIssuers();
+        int index = 0;
+        while(certList.size() > index) {
+            System.out.println("Certificate issuer: " + certList.get(index).getIssuerX500Principal());
+            for (X509Certificate validIssuer : validIssuers) {
+                if (certList.get(index).getIssuerX500Principal().equals(validIssuer.getIssuerX500Principal())) {
+                    System.out.println("Found a valid issuer, checking sig");
+                    //verify that the signature from the root is signed by the valid authority
+                    validIssuer.verify(certList.get(certList.size() - 1).getPublicKey());
+                    System.out.println("Signature verification passed");
+                    return validIssuer;
+                }
+            }
+            index++;
+        }
+        System.out.println("Valid issuer not found");
+        return null;
+    }
+
+    private static List<X509Certificate> selectKeyData(List list)
+    {
+        PublicKey recovered = null;
+        X509Certificate cert = null;
+        List<X509Certificate> certList = new ArrayList<>();
+
+        for (int i = 0; i < list.size(); i++) {
+            XMLStructure xmlStructure = (XMLStructure) list.get(i);
+
+            if (xmlStructure instanceof KeyValue)
+            {
+                KeyValue kv = (KeyValue)xmlStructure;
+                try
+                {
+                    recovered = kv.getPublicKey();
+                }
+                catch (KeyException e)
+                {
+                    e.printStackTrace();
+                }
+            }
+            if (xmlStructure instanceof X509Data) {
+                List certs = ((X509Data) xmlStructure).getContent();
+                certList.addAll(certs);
+                cert = certList.get(certList.size() - 1);
+                for (X509Certificate crt : certList)
+                {
+                    try
+                    {
+                        crt.checkValidity();
+                        if (recovered != null)
+                        {
+                            PublicKey certKey = crt.getPublicKey();
+                            if (Arrays.equals(recovered.getEncoded(), certKey.getEncoded()))
+                            {
+                                cert = crt;
+                            }
+                        }
+                        else
+                        {
+                            if (crt.getSigAlgName().equals("SHA256withECDSA"))
+                            {
+                                cert = crt;
```

Comment:
Cannot recover the public key unless it is RSA? @JamesSmartCell @colourful-land

Reply:
I do not know what the quoted code intends, but speaking from the point of Java Cryptography Architecture, it's clearly possible to get the EC public key from a cert. Just add this to see for yourself:

              169      try
              170      {
              171          crt.checkValidity();
              172          System.out.println(crt.getPublicKey().toString());

That outputs this:

    io.stormbird.token.tools.TokenScriptValidatorTest > verifyECDSAxmldsig STANDARD_OUT
    Sun EC public key, 256 bits
      public x coord: 24544810233457759196143703211090909005083847164792591927755581006921757564400
      public y coord: 53137840816744055503003735497360707636898773073824264522122612147305425927261
      parameters: secp256r1 [NIST P-256, X9.62 prime256v1] (1.2.840.10045.3.1.7)
    Sun EC public key, 256 bits
      public x coord: 24544810233457759196143703211090909005083847164792591927755581006921757564400
      public y coord: 53137840816744055503003735497360707636898773073824264522122612147305425927261
      parameters: secp256r1 [NIST P-256, X9.62 prime256v1] (1.2.840.10045.3.1.7)

(it outputs 2 times because I wasn't using your post-refactor code. @JamesSmartCell 's old code goes through certs 2 times)

Reply:
This is making me more confused. Why recover the public key if it is going to checkValidity on the cert? My refactoring has made this function fail but I am having trouble pinpointing the issue. Still looking.